### PR TITLE
DisposableScheduler

### PR DESCRIPTION
**Pull request by FrankBakkerNl**
## Proposed change

Support Cron scheduling as an Extension of the IScheduler interface and allow 
injection of an System.Reactive.Concurrency.IScheduler that logs exceptions and cancels scheduled tasks when the app is stopped.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code compiles without warnings (code quality chek)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/net-daemon/docs

### CODE REVIEW

File: src/Extensions/NetDaemon.Extensions.Scheduling.Tests/Scheduling/DisposableSchedulerTest.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+        inner.AdvanceBy(TimeSpan.FromMinutes(1).Ticks);
+        called.Should().Be(2);
+
+        disposable.Dispose();
+        inner.AdvanceBy(TimeSpan.FromMinutes(1).Ticks);
+        called.Should().Be(2);
+    }
+    
+    [Fact]
```
Comment (FrankBakkerNl) [inline src/Extensions/NetDaemon.Extensions.Scheduling.Tests/Scheduling/DisposableSchedulerTest.cs:34]: This test fails to stop the sheduler

File: src/Extensions/NetDaemon.Extensions.Scheduling.Tests/Scheduling/DisposableSchedulerTest.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        var sub = disposable.SchedulePeriodic(TimeSpan.FromMinutes(1), () => called++);
+        inner.AdvanceBy(TimeSpan.FromMinutes(1).Ticks);
+        called.Should().Be(1);
+
+        inner.AdvanceBy(TimeSpan.FromMinutes(1).Ticks);
+        called.Should().Be(2);
+        sub.Dispose();
+
+        inner.AdvanceBy(TimeSpan.FromMinutes(1).Ticks);
+        called.Should().Be(2);
```
Comment (helto4real) [inline src/Extensions/NetDaemon.Extensions.Scheduling.Tests/Scheduling/DisposableSchedulerTest.cs:52]:
```c#
called.Should().Be(3);
```

File: src/Extensions/NetDaemon.Extensions.Scheduling.Tests/Scheduling/DisposableSchedulerTest.cs (hunk 1 above)
Comment (helto4real) [reply src/Extensions/NetDaemon.Extensions.Scheduling.Tests/Scheduling/DisposableSchedulerTest.cs:34]: The test look valid though

File: src/Extensions/NetDaemon.Extensions.Scheduling/DisposableScheduler.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    /// <inheritdoc />
+    public IDisposable Schedule<TState>(TState state, DateTimeOffset dueTime, Func<IScheduler, TState, IDisposable> action)
+    {
+        if (_isDisposed) return Disposable.Empty;
+
+        CancellationTokenRegistration? cancellationTokenRegistration = null;
+
+        var subscription = _innerScheduler.Schedule(state, dueTime, (s, d) =>
+        {
+            cancellationTokenRegistration?.Unregister();
```
Comment (FrankBakkerNl) [inline src/Extensions/NetDaemon.Extensions.Scheduling/DisposableScheduler.cs:61]: This bit feels a bit weird, I need to capture the cancellationTokenRegistration in the delagate in order to unregister it, but at this point it is not definitively assigned yet

File: src/Extensions/NetDaemon.Extensions.Scheduling/SchedulerExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    /// <inheritdoc/>
+    public static IDisposable RunEvery(IScheduler scheduler, TimeSpan period, DateTimeOffset startTime, Action action)
+    {
+        return Observable.Timer(startTime, period, scheduler).Subscribe(_ => action());
+    }
+
+    /// <inheritdoc/>
+    public static IDisposable RunIn(IScheduler scheduler, TimeSpan timespan, Action action)
```
Comment (FrankBakkerNl) [inline src/Extensions/NetDaemon.Extensions.Scheduling/SchedulerExtensions.cs:22]: These are not very usefull anymore, they are just different names for something that already exists

File: src/Extensions/NetDaemon.Extensions.Scheduling/SchedulerExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+{
+    /// <summary>
+    ///     Schedules an action every (timespan)
+    /// </summary>
+    /// <param name="scheduler">The Scheduler to use</param>
+    /// <param name="timespan">The period to schedule</param>
+    /// <param name="action">Action to run</param>
+    public static IDisposable RunEvery(this IScheduler scheduler, TimeSpan timespan, Action action)
+    {
+        return scheduler.SchedulePeriodic(timespan, action);
```
Comment (FrankBakkerNl) [inline src/Extensions/NetDaemon.Extensions.Scheduling/SchedulerExtensions.cs:21]: Most of these are not very usefull anymore, they are just different names for things that already exists

Comment (helto4real) [reply src/Extensions/NetDaemon.Extensions.Scheduling/SchedulerExtensions.cs:21]: Agree we should just use the default. This is a new interface anyway. The important thing is the error handling safety.

File: src/Extensions/NetDaemon.Extensions.Scheduling/DependencyInjectionSetup.cs [hunk 6]
```diff
@@ -14,6 +16,7 @@ public static class DependencyInjectionSetup
     public static IServiceCollection AddNetDaemonScheduler(this IServiceCollection services)
     {
         services.AddScoped<INetDaemonScheduler, NetDaemonScheduler>();
+        services.AddScoped<IScheduler>(s => new DisposableScheduler(DefaultScheduler.Instance.WrapWithLogger(s.GetRequiredService<ILogger<IScheduler>>())));
```
Comment (helto4real) [inline src/Extensions/NetDaemon.Extensions.Scheduling/DependencyInjectionSetup.cs:19]: this is much nicer solution than mine was :)

**Review by helto4real — APPROVED** [review]

**Pull request merged**